net/WebSocketHandler.hpp

```diff
@@ -543,54 +543,66 @@ class WebSocketHandler : public ProtocolHandlerInterface
+    void buildFrame(const char* data, const uint64_t len, unsigned char flags, Buffer &out) const
     {
-        out.push_back(flags);
+        int slen = 0;
+        char scratch[16];
+
+        scratch[slen++] = flags;
 
         int maskFlag = _isMasking ? 0x80 : 0;
         if (len < 126)
         {
-            out.push_back((char)(len | maskFlag));
+            scratch[slen++] = (char)(len | maskFlag);
         }
         else if (len <= 0xffff)
         {
-            out.push_back((char)(126 | maskFlag));
-            out.push_back(static_cast<char>((len >> 8) & 0xff));
-            out.push_back(static_cast<char>((len >> 0) & 0xff));
+            scratch[slen++] = (char)(126 | maskFlag);
+            scratch[slen++] = static_cast<char>((len >> 8) & 0xff);
+            scratch[slen++] = static_cast<char>((len >> 0) & 0xff);
         }
         else
         {
-            out.push_back((char)(127 | maskFlag));
-            out.push_back(static_cast<char>((len >> 56) & 0xff));
-            out.push_back(static_cast<char>((len >> 48) & 0xff));
-            out.push_back(static_cast<char>((len >> 40) & 0xff));
-            out.push_back(static_cast<char>((len >> 32) & 0xff));
-            out.push_back(static_cast<char>((len >> 24) & 0xff));
-            out.push_back(static_cast<char>((len >> 16) & 0xff));
-            out.push_back(static_cast<char>((len >> 8) & 0xff));
-            out.push_back(static_cast<char>((len >> 0) & 0xff));
+            scratch[slen++] = (char)(127 | maskFlag);
+            scratch[slen++] = static_cast<char>((len >> 56) & 0xff);
+            scratch[slen++] = static_cast<char>((len >> 48) & 0xff);
+            scratch[slen++] = static_cast<char>((len >> 40) & 0xff);
+            scratch[slen++] = static_cast<char>((len >> 32) & 0xff);
+            scratch[slen++] = static_cast<char>((len >> 24) & 0xff);
+            scratch[slen++] = static_cast<char>((len >> 16) & 0xff);
+            scratch[slen++] = static_cast<char>((len >> 8) & 0xff);
+            scratch[slen++] = static_cast<char>((len >> 0) & 0xff);
         }
 
+        out.append(scratch, slen);
+
         if (_isMasking)
         { // flip some top bits - perhaps it helps.
-            size_t mask = out.size();
-
-            out.push_back(static_cast<char>(0x81));
-            out.push_back(static_cast<char>(0x76));
-            out.push_back(static_cast<char>(0x81));
-            out.push_back(static_cast<char>(0x76));
-
-            // Copy the data.
-            out.insert(out.end(), data, data + len);
-
-            // Mask it.
-            for (size_t i = 4; i < out.size() - mask; ++i)
-                out[mask + i] = out[mask + i] ^ out[mask + (i%4)];
+            char mask[4];
+
+            mask[0] = static_cast<char>(0x81);
+            mask[1] = static_cast<char>(0x76);
+            mask[2] = static_cast<char>(0x81);
+            mask[3] = static_cast<char>(0x76);
+            out.append(scratch, 4);
```

Comment:
Shouldn't this be `mask` instead of `scratch`?

Reply:
Yes - amazing that it works without that; hmm - I wonder why that is ? perhaps we're not doing internal masking anymore between Kit and WSD, and we don't do it outbound (only inbound) - not sure; interesting indeed =) good to know that's not a hot path.

Reply:
Internally it's disabled. It seems to be only enabled on the Admin connection. Technically, only clients need to mask, if they choose to, and loolwsd typically isn't, not even for admin (so not sure why it's enabled there).